libraries/chain/include/eosio/chain/controller.hpp

```diff
@@ -394,14 +388,12 @@ namespace eosio { namespace chain {
          void replace_producer_keys( const public_key_type& key );
          void replace_account_keys( name account, name permission, const public_key_type& key );
 
-         eosio::chain::combined_database& kv_db()const;
-
+      chainbase::database& mutable_db()const;
+      
```

Comment:
I would prefer for this to remain private. However, I don't like having to forward declare `chain_apis` namespace in order to add a friend. Best thing would be to create a const kv_context, however that also gets rather ugly because of the need for a const iterator etc. I think you can change `create_kv_context` to take a controller and then add it as a friend.

Reply:
did it. friending other namespace function appears to be pretty verbose.